# DATA PREPROCESSING

In [62]:
import pandas as pd

data = pd.read_csv('Reviews.csv', nrows=10000)

data

Id   ProductId          UserId                      ProfileName  \
0         1  B001E4KFG0  A3SGXH7AUHU8GW                       delmartian   
1         2  B00813GRG4  A1D87F6ZCVE5NK                           dll pa   
2         3  B000LQOCH0   ABXLMWJIXXAIN  Natalia Corres "Natalia Corres"   
3         4  B000UA0QIQ  A395BORC6FGVXV                             Karl   
4         5  B006K2ZZ7K  A1UQRSCLF8GW1T    Michael D. Bigham "M. Wassir"   
...     ...         ...             ...                              ...   
9995   9996  B000P41A28  A3A63RACXR1XIL            A. Boodhoo "deaddodo"   
9996   9997  B000P41A28    A5VVRGL8JA7R                             Adam   
9997   9998  B000P41A28  A2TGDTJ8YCU6PD                          geena77   
9998   9999  B000P41A28   AUV4GIZZE693O              Susan Coe "sueysis"   
9999  10000  B000P41A28   A82WIMR4RSVLI                       Emrose mom   

      HelpfulnessNumerator  HelpfulnessDenominator  Score        Time  \
0                        1                       1      5  1303862400   
1                        0                       0      1  1346976000   
2                        1                       1      4  1219017600   
3                        3                       3      2  1307923200   
4                        0                       0      5  1350777600   
...                    ...                     ...    ...         ...   
9995                    10                      15      1  1204502400   
9996                     2                       3      5  1306368000   
9997                     0                       0      5  1347494400   
9998                     1                       2      5  1203638400   
9999                     0                       1      4  1337472000   

                               Summary  \
0                Good Quality Dog Food   
1                    Not as Advertised   
2                "Delight" says it all   
3                       Cough Medicine   
4                          Great taffy   
...                                ...   
9995                      constipation   
9996  Constipation Not A Problem if...   
9997                Love this formula!   
9998                   very convenient   
9999        The best weve tried so far   

                                                   Text  
0     I have bought several of the Vitality canned d...  
1     Product arrived labeled as Jumbo Salted Peanut...  
2     This is a confection that has been around a fe...  
3     If you are looking for the secret ingredient i...  
4     Great taffy at a great price.  There was a wid...  
...                                                 ...  
9995  we switched from the advance similac to the or...  
9996  Like the bad reviews say, the organic formula ...  
9997  I wanted to solely breastfeed but was unable t...  
9998  i love the fact that i can get this delieved t...  
9999  We have a 7 week old... He had gas and constip...  

[10000 rows x 10 columns]

In [63]:
import string

def remove_punctuation(text):
    punctuation_free = ""
    for i in text:
        if i not in string.punctuation:
            punctuation_free += i
    return punctuation_free
data['Clean Punctuation']= data['Text'].apply(remove_punctuation)
data

Id   ProductId          UserId                      ProfileName  \
0         1  B001E4KFG0  A3SGXH7AUHU8GW                       delmartian   
1         2  B00813GRG4  A1D87F6ZCVE5NK                           dll pa   
2         3  B000LQOCH0   ABXLMWJIXXAIN  Natalia Corres "Natalia Corres"   
3         4  B000UA0QIQ  A395BORC6FGVXV                             Karl   
4         5  B006K2ZZ7K  A1UQRSCLF8GW1T    Michael D. Bigham "M. Wassir"   
...     ...         ...             ...                              ...   
9995   9996  B000P41A28  A3A63RACXR1XIL            A. Boodhoo "deaddodo"   
9996   9997  B000P41A28    A5VVRGL8JA7R                             Adam   
9997   9998  B000P41A28  A2TGDTJ8YCU6PD                          geena77   
9998   9999  B000P41A28   AUV4GIZZE693O              Susan Coe "sueysis"   
9999  10000  B000P41A28   A82WIMR4RSVLI                       Emrose mom   

      HelpfulnessNumerator  HelpfulnessDenominator  Score        Time  \
0                        1                       1      5  1303862400   
1                        0                       0      1  1346976000   
2                        1                       1      4  1219017600   
3                        3                       3      2  1307923200   
4                        0                       0      5  1350777600   
...                    ...                     ...    ...         ...   
9995                    10                      15      1  1204502400   
9996                     2                       3      5  1306368000   
9997                     0                       0      5  1347494400   
9998                     1                       2      5  1203638400   
9999                     0                       1      4  1337472000   

                               Summary  \
0                Good Quality Dog Food   
1                    Not as Advertised   
2                "Delight" says it all   
3                       Cough Medicine   
4                          Great taffy   
...                                ...   
9995                      constipation   
9996  Constipation Not A Problem if...   
9997                Love this formula!   
9998                   very convenient   
9999        The best weve tried so far   

                                                   Text  \
0     I have bought several of the Vitality canned d...   
1     Product arrived labeled as Jumbo Salted Peanut...   
2     This is a confection that has been around a fe...   
3     If you are looking for the secret ingredient i...   
4     Great taffy at a great price.  There was a wid...   
...                                                 ...   
9995  we switched from the advance similac to the or...   
9996  Like the bad reviews say, the organic formula ...   
9997  I wanted to solely breastfeed but was unable t...   
9998  i love the fact that i can get this delieved t...   
9999  We have a 7 week old... He had gas and constip...   

                                      Clean Punctuation  
0     I have bought several of the Vitality canned d...  
1     Product arrived labeled as Jumbo Salted Peanut...  
2     This is a confection that has been around a fe...  
3     If you are looking for the secret ingredient i...  
4     Great taffy at a great price  There was a wide...  
...                                                 ...  
9995  we switched from the advance similac to the or...  
9996  Like the bad reviews say the organic formula c...  
9997  I wanted to solely breastfeed but was unable t...  
9998  i love the fact that i can get this delieved t...  
9999  We have a 7 week old He had gas and constipati...  

[10000 rows x 11 columns]

In [64]:
data['Clean Lowercase']= data['Clean Punctuation'].str.lower()
data 


Id   ProductId          UserId                      ProfileName  \
0         1  B001E4KFG0  A3SGXH7AUHU8GW                       delmartian   
1         2  B00813GRG4  A1D87F6ZCVE5NK                           dll pa   
2         3  B000LQOCH0   ABXLMWJIXXAIN  Natalia Corres "Natalia Corres"   
3         4  B000UA0QIQ  A395BORC6FGVXV                             Karl   
4         5  B006K2ZZ7K  A1UQRSCLF8GW1T    Michael D. Bigham "M. Wassir"   
...     ...         ...             ...                              ...   
9995   9996  B000P41A28  A3A63RACXR1XIL            A. Boodhoo "deaddodo"   
9996   9997  B000P41A28    A5VVRGL8JA7R                             Adam   
9997   9998  B000P41A28  A2TGDTJ8YCU6PD                          geena77   
9998   9999  B000P41A28   AUV4GIZZE693O              Susan Coe "sueysis"   
9999  10000  B000P41A28   A82WIMR4RSVLI                       Emrose mom   

      HelpfulnessNumerator  HelpfulnessDenominator  Score        Time  \
0                        1                       1      5  1303862400   
1                        0                       0      1  1346976000   
2                        1                       1      4  1219017600   
3                        3                       3      2  1307923200   
4                        0                       0      5  1350777600   
...                    ...                     ...    ...         ...   
9995                    10                      15      1  1204502400   
9996                     2                       3      5  1306368000   
9997                     0                       0      5  1347494400   
9998                     1                       2      5  1203638400   
9999                     0                       1      4  1337472000   

                               Summary  \
0                Good Quality Dog Food   
1                    Not as Advertised   
2                "Delight" says it all   
3                       Cough Medicine   
4                          Great taffy   
...                                ...   
9995                      constipation   
9996  Constipation Not A Problem if...   
9997                Love this formula!   
9998                   very convenient   
9999        The best weve tried so far   

                                                   Text  \
0     I have bought several of the Vitality canned d...   
1     Product arrived labeled as Jumbo Salted Peanut...   
2     This is a confection that has been around a fe...   
3     If you are looking for the secret ingredient i...   
4     Great taffy at a great price.  There was a wid...   
...                                                 ...   
9995  we switched from the advance similac to the or...   
9996  Like the bad reviews say, the organic formula ...   
9997  I wanted to solely breastfeed but was unable t...   
9998  i love the fact that i can get this delieved t...   
9999  We have a 7 week old... He had gas and constip...   

                                      Clean Punctuation  \
0     I have bought several of the Vitality canned d...   
1     Product arrived labeled as Jumbo Salted Peanut...   
2     This is a confection that has been around a fe...   
3     If you are looking for the secret ingredient i...   
4     Great taffy at a great price  There was a wide...   
...                                                 ...   
9995  we switched from the advance similac to the or...   
9996  Like the bad reviews say the organic formula c...   
9997  I wanted to solely breastfeed but was unable t...   
9998  i love the fact that i can get this delieved t...   
9999  We have a 7 week old He had gas and constipati...   

                                        Clean Lowercase  
0     i have bought several of the vitality canned d...  
1     product arrived labeled as jumbo salted peanut...  
2     this is a confection that has been around a fe...  
3     if you are looking for the secret ingredient i...  
4     

In [65]:
import re

def remove_numbers(text):
    return re.sub("[\d-]",'',text)
data['Clean Number'] = data['Clean Lowercase'].apply(remove_numbers)
data


Id   ProductId          UserId                      ProfileName  \
0         1  B001E4KFG0  A3SGXH7AUHU8GW                       delmartian   
1         2  B00813GRG4  A1D87F6ZCVE5NK                           dll pa   
2         3  B000LQOCH0   ABXLMWJIXXAIN  Natalia Corres "Natalia Corres"   
3         4  B000UA0QIQ  A395BORC6FGVXV                             Karl   
4         5  B006K2ZZ7K  A1UQRSCLF8GW1T    Michael D. Bigham "M. Wassir"   
...     ...         ...             ...                              ...   
9995   9996  B000P41A28  A3A63RACXR1XIL            A. Boodhoo "deaddodo"   
9996   9997  B000P41A28    A5VVRGL8JA7R                             Adam   
9997   9998  B000P41A28  A2TGDTJ8YCU6PD                          geena77   
9998   9999  B000P41A28   AUV4GIZZE693O              Susan Coe "sueysis"   
9999  10000  B000P41A28   A82WIMR4RSVLI                       Emrose mom   

      HelpfulnessNumerator  HelpfulnessDenominator  Score        Time  \
0                        1                       1      5  1303862400   
1                        0                       0      1  1346976000   
2                        1                       1      4  1219017600   
3                        3                       3      2  1307923200   
4                        0                       0      5  1350777600   
...                    ...                     ...    ...         ...   
9995                    10                      15      1  1204502400   
9996                     2                       3      5  1306368000   
9997                     0                       0      5  1347494400   
9998                     1                       2      5  1203638400   
9999                     0                       1      4  1337472000   

                               Summary  \
0                Good Quality Dog Food   
1                    Not as Advertised   
2                "Delight" says it all   
3                       Cough Medicine   
4                          Great taffy   
...                                ...   
9995                      constipation   
9996  Constipation Not A Problem if...   
9997                Love this formula!   
9998                   very convenient   
9999        The best weve tried so far   

                                                   Text  \
0     I have bought several of the Vitality canned d...   
1     Product arrived labeled as Jumbo Salted Peanut...   
2     This is a confection that has been around a fe...   
3     If you are looking for the secret ingredient i...   
4     Great taffy at a great price.  There was a wid...   
...                                                 ...   
9995  we switched from the advance similac to the or...   
9996  Like the bad reviews say, the organic formula ...   
9997  I wanted to solely breastfeed but was unable t...   
9998  i love the fact that i can get this delieved t...   
9999  We have a 7 week old... He had gas and constip...   

                                      Clean Punctuation  \
0     I have bought several of the Vitality canned d...   
1     Product arrived labeled as Jumbo Salted Peanut...   
2     This is a confection that has been around a fe...   
3     If you are looking for the secret ingredient i...   
4     Great taffy at a great price  There was a wide...   
...                                                 ...   
9995  we switched from the advance similac to the or...   
9996  Like the bad reviews say the organic formula c...   
9997  I wanted to solely breastfeed but was unable t...   
9998  i love the fact that i can get this delieved t...   
9999  We have a 7 week old He had gas and constipati...   

                                        Clean Lowercase  \
0     i have bought several of the vitality canned d...   
1     product arrived labeled as jumbo salted peanut...   
2     this is a confection that has been around a fe...   
3     if you are looking for the secret ingredient i...   
4

In [66]:
import nltk
nltk.download('punkt')
from nltk.tokenize import word_tokenize
data['Token Data']= data['Clean Number'].apply(word_tokenize)
data

[nltk_data] Downloading package punkt to
[nltk_data]     /Users/zafirahzaini/nltk_data...
[nltk_data]   Package punkt is already up-to-date!


Id   ProductId          UserId                      ProfileName  \
0         1  B001E4KFG0  A3SGXH7AUHU8GW                       delmartian   
1         2  B00813GRG4  A1D87F6ZCVE5NK                           dll pa   
2         3  B000LQOCH0   ABXLMWJIXXAIN  Natalia Corres "Natalia Corres"   
3         4  B000UA0QIQ  A395BORC6FGVXV                             Karl   
4         5  B006K2ZZ7K  A1UQRSCLF8GW1T    Michael D. Bigham "M. Wassir"   
...     ...         ...             ...                              ...   
9995   9996  B000P41A28  A3A63RACXR1XIL            A. Boodhoo "deaddodo"   
9996   9997  B000P41A28    A5VVRGL8JA7R                             Adam   
9997   9998  B000P41A28  A2TGDTJ8YCU6PD                          geena77   
9998   9999  B000P41A28   AUV4GIZZE693O              Susan Coe "sueysis"   
9999  10000  B000P41A28   A82WIMR4RSVLI                       Emrose mom   

      HelpfulnessNumerator  HelpfulnessDenominator  Score        Time  \
0                        1                       1      5  1303862400   
1                        0                       0      1  1346976000   
2                        1                       1      4  1219017600   
3                        3                       3      2  1307923200   
4                        0                       0      5  1350777600   
...                    ...                     ...    ...         ...   
9995                    10                      15      1  1204502400   
9996                     2                       3      5  1306368000   
9997                     0                       0      5  1347494400   
9998                     1                       2      5  1203638400   
9999                     0                       1      4  1337472000   

                               Summary  \
0                Good Quality Dog Food   
1                    Not as Advertised   
2                "Delight" says it all   
3                       Cough Medicine   
4                          Great taffy   
...                                ...   
9995                      constipation   
9996  Constipation Not A Problem if...   
9997                Love this formula!   
9998                   very convenient   
9999        The best weve tried so far   

                                                   Text  \
0     I have bought several of the Vitality canned d...   
1     Product arrived labeled as Jumbo Salted Peanut...   
2     This is a confection that has been around a fe...   
3     If you are looking for the secret ingredient i...   
4     Great taffy at a great price.  There was a wid...   
...                                                 ...   
9995  we switched from the advance similac to the or...   
9996  Like the bad reviews say, the organic formula ...   
9997  I wanted to solely breastfeed but was unable t...   
9998  i love the fact that i can get this delieved t...   
9999  We have a 7 week old... He had gas and constip...   

                                      Clean Punctuation  \
0     I have bought several of the Vitality canned d...   
1     Product arrived labeled as Jumbo Salted Peanut...   
2     This is a confection that has been around a fe...   
3     If you are looking for the secret ingredient i...   
4     Great taffy at a great price  There was a wide...   
...                                                 ...   
9995  we switched from the advance similac to the or...   
9996  Like the bad reviews say the organic formula c...   
9997  I wanted to solely breastfeed but was unable t...   
9998  i love the fact that i can get this delieved t...   
9999  We have a 7 week old He had gas and constipati...   

                                        Clean Lowercase  \
0     i have bought several of the vitality canned d...   
1     product arrived labeled as jumbo salted peanut...   
2     this is a confection that has been around a fe...   
3     if you are looking for the secret ingredient i...   
4

In [67]:
nltk.download('stopwords')
stopwords = nltk.corpus.stopwords.words('english')
print(stopwords)


['i', 'me', 'my', 'myself', 'we', 'our', 'ours', 'ourselves', 'you', "you're", "you've", "you'll", "you'd", 'your', 'yours', 'yourself', 'yourselves', 'he', 'him', 'his', 'himself', 'she', "she's", 'her', 'hers', 'herself', 'it', "it's", 'its', 'itself', 'they', 'them', 'their', 'theirs', 'themselves', 'what', 'which', 'who', 'whom', 'this', 'that', "that'll", 'these', 'those', 'am', 'is', 'are', 'was', 'were', 'be', 'been', 'being', 'have', 'has', 'had', 'having', 'do', 'does', 'did', 'doing', 'a', 'an', 'the', 'and', 'but', 'if', 'or', 'because', 'as', 'until', 'while', 'of', 'at', 'by', 'for', 'with', 'about', 'against', 'between', 'into', 'through', 'during', 'before', 'after', 'above', 'below', 'to', 'from', 'up', 'down', 'in', 'out', 'on', 'off', 'over', 'under', 'again', 'further', 'then', 'once', 'here', 'there', 'when', 'where', 'why', 'how', 'all', 'any', 'both', 'each', 'few', 'more', 'most', 'other', 'some', 'such', 'no', 'nor', 'not', 'only', 'own', 'same', 'so', 'than', '

[nltk_data] Downloading package stopwords to
[nltk_data]     /Users/zafirahzaini/nltk_data...
[nltk_data]   Package stopwords is already up-to-date!


In [68]:
def remove_stopwords(text):
    output = []
    for i in text:
        if i not in stopwords:
            output.append(i)
    return output
data['Clean Stopwords'] = data['Token Data'].apply(remove_stopwords)
data

Id   ProductId          UserId                      ProfileName  \
0         1  B001E4KFG0  A3SGXH7AUHU8GW                       delmartian   
1         2  B00813GRG4  A1D87F6ZCVE5NK                           dll pa   
2         3  B000LQOCH0   ABXLMWJIXXAIN  Natalia Corres "Natalia Corres"   
3         4  B000UA0QIQ  A395BORC6FGVXV                             Karl   
4         5  B006K2ZZ7K  A1UQRSCLF8GW1T    Michael D. Bigham "M. Wassir"   
...     ...         ...             ...                              ...   
9995   9996  B000P41A28  A3A63RACXR1XIL            A. Boodhoo "deaddodo"   
9996   9997  B000P41A28    A5VVRGL8JA7R                             Adam   
9997   9998  B000P41A28  A2TGDTJ8YCU6PD                          geena77   
9998   9999  B000P41A28   AUV4GIZZE693O              Susan Coe "sueysis"   
9999  10000  B000P41A28   A82WIMR4RSVLI                       Emrose mom   

      HelpfulnessNumerator  HelpfulnessDenominator  Score        Time  \
0                        1                       1      5  1303862400   
1                        0                       0      1  1346976000   
2                        1                       1      4  1219017600   
3                        3                       3      2  1307923200   
4                        0                       0      5  1350777600   
...                    ...                     ...    ...         ...   
9995                    10                      15      1  1204502400   
9996                     2                       3      5  1306368000   
9997                     0                       0      5  1347494400   
9998                     1                       2      5  1203638400   
9999                     0                       1      4  1337472000   

                               Summary  \
0                Good Quality Dog Food   
1                    Not as Advertised   
2                "Delight" says it all   
3                       Cough Medicine   
4                          Great taffy   
...                                ...   
9995                      constipation   
9996  Constipation Not A Problem if...   
9997                Love this formula!   
9998                   very convenient   
9999        The best weve tried so far   

                                                   Text  \
0     I have bought several of the Vitality canned d...   
1     Product arrived labeled as Jumbo Salted Peanut...   
2     This is a confection that has been around a fe...   
3     If you are looking for the secret ingredient i...   
4     Great taffy at a great price.  There was a wid...   
...                                                 ...   
9995  we switched from the advance similac to the or...   
9996  Like the bad reviews say, the organic formula ...   
9997  I wanted to solely breastfeed but was unable t...   
9998  i love the fact that i can get this delieved t...   
9999  We have a 7 week old... He had gas and constip...   

                                      Clean Punctuation  \
0     I have bought several of the Vitality canned d...   
1     Product arrived labeled as Jumbo Salted Peanut...   
2     This is a confection that has been around a fe...   
3     If you are looking for the secret ingredient i...   
4     Great taffy at a great price  There was a wide...   
...                                                 ...   
9995  we switched from the advance similac to the or...   
9996  Like the bad reviews say the organic formula c...   
9997  I wanted to solely breastfeed but was unable t...   
9998  i love the fact that i can get this delieved t...   
9999  We have a 7 week old He had gas and constipati...   

                                        Clean Lowercase  \
0     i have bought several of the vitality canned d...   
1     product arrived labeled as jumbo salted peanut...   
2     this is a confection that has been around a fe...   
3     if you are looking for the secret ingredient i...   
4

In [69]:
from nltk.stem.porter import PorterStemmer
porter_stemmer = PorterStemmer()

def stemming(text):
    stem_text = []
    for word in text:
        stemmed_word = porter_stemmer.stem(word)
        stem_text.append(stemmed_word)
    return stem_text

data['Clean Stemmed'] = data['Clean Stopwords'].apply(stemming)
data


Id   ProductId          UserId                      ProfileName  \
0         1  B001E4KFG0  A3SGXH7AUHU8GW                       delmartian   
1         2  B00813GRG4  A1D87F6ZCVE5NK                           dll pa   
2         3  B000LQOCH0   ABXLMWJIXXAIN  Natalia Corres "Natalia Corres"   
3         4  B000UA0QIQ  A395BORC6FGVXV                             Karl   
4         5  B006K2ZZ7K  A1UQRSCLF8GW1T    Michael D. Bigham "M. Wassir"   
...     ...         ...             ...                              ...   
9995   9996  B000P41A28  A3A63RACXR1XIL            A. Boodhoo "deaddodo"   
9996   9997  B000P41A28    A5VVRGL8JA7R                             Adam   
9997   9998  B000P41A28  A2TGDTJ8YCU6PD                          geena77   
9998   9999  B000P41A28   AUV4GIZZE693O              Susan Coe "sueysis"   
9999  10000  B000P41A28   A82WIMR4RSVLI                       Emrose mom   

      HelpfulnessNumerator  HelpfulnessDenominator  Score        Time  \
0                        1                       1      5  1303862400   
1                        0                       0      1  1346976000   
2                        1                       1      4  1219017600   
3                        3                       3      2  1307923200   
4                        0                       0      5  1350777600   
...                    ...                     ...    ...         ...   
9995                    10                      15      1  1204502400   
9996                     2                       3      5  1306368000   
9997                     0                       0      5  1347494400   
9998                     1                       2      5  1203638400   
9999                     0                       1      4  1337472000   

                               Summary  \
0                Good Quality Dog Food   
1                    Not as Advertised   
2                "Delight" says it all   
3                       Cough Medicine   
4                          Great taffy   
...                                ...   
9995                      constipation   
9996  Constipation Not A Problem if...   
9997                Love this formula!   
9998                   very convenient   
9999        The best weve tried so far   

                                                   Text  \
0     I have bought several of the Vitality canned d...   
1     Product arrived labeled as Jumbo Salted Peanut...   
2     This is a confection that has been around a fe...   
3     If you are looking for the secret ingredient i...   
4     Great taffy at a great price.  There was a wid...   
...                                                 ...   
9995  we switched from the advance similac to the or...   
9996  Like the bad reviews say, the organic formula ...   
9997  I wanted to solely breastfeed but was unable t...   
9998  i love the fact that i can get this delieved t...   
9999  We have a 7 week old... He had gas and constip...   

                                      Clean Punctuation  \
0     I have bought several of the Vitality canned d...   
1     Product arrived labeled as Jumbo Salted Peanut...   
2     This is a confection that has been around a fe...   
3     If you are looking for the secret ingredient i...   
4     Great taffy at a great price  There was a wide...   
...                                                 ...   
9995  we switched from the advance similac to the or...   
9996  Like the bad reviews say the organic formula c...   
9997  I wanted to solely breastfeed but was unable t...   
9998  i love the fact that i can get this delieved t...   
9999  We have a 7 week old He had gas and constipati...   

                                        Clean Lowercase  \
0     i have bought several of the vitality canned d...   
1     product arrived labeled as jumbo salted peanut...   
2     this is a confection that has been around a fe...   
3     if you are looking for the secret ingredient i...   
4

In [72]:
nltk.download('wordnet')

from nltk.stem import WordNetLemmatizer
wordnet_lemmatizer = WordNetLemmatizer()

def lemmatizer(text):
    lemm_text = []
    for word in text:
        lemmatized_word = wordnet_lemmatizer.lemmatize(word)
        lemm_text.append(lemmatized_word)
    return lemm_text

data['Clean Lemmatized1']=data['Clean Stopwords'].apply(lemmatizer)

data['Clean Lemmatized2']=data['Clean Stemmed'].apply(lemmatizer)
data


[nltk_data] Downloading package wordnet to
[nltk_data]     /Users/zafirahzaini/nltk_data...
[nltk_data]   Package wordnet is already up-to-date!


Id   ProductId          UserId                      ProfileName  \
0         1  B001E4KFG0  A3SGXH7AUHU8GW                       delmartian   
1         2  B00813GRG4  A1D87F6ZCVE5NK                           dll pa   
2         3  B000LQOCH0   ABXLMWJIXXAIN  Natalia Corres "Natalia Corres"   
3         4  B000UA0QIQ  A395BORC6FGVXV                             Karl   
4         5  B006K2ZZ7K  A1UQRSCLF8GW1T    Michael D. Bigham "M. Wassir"   
...     ...         ...             ...                              ...   
9995   9996  B000P41A28  A3A63RACXR1XIL            A. Boodhoo "deaddodo"   
9996   9997  B000P41A28    A5VVRGL8JA7R                             Adam   
9997   9998  B000P41A28  A2TGDTJ8YCU6PD                          geena77   
9998   9999  B000P41A28   AUV4GIZZE693O              Susan Coe "sueysis"   
9999  10000  B000P41A28   A82WIMR4RSVLI                       Emrose mom   

      HelpfulnessNumerator  HelpfulnessDenominator  Score        Time  \
0                        1                       1      5  1303862400   
1                        0                       0      1  1346976000   
2                        1                       1      4  1219017600   
3                        3                       3      2  1307923200   
4                        0                       0      5  1350777600   
...                    ...                     ...    ...         ...   
9995                    10                      15      1  1204502400   
9996                     2                       3      5  1306368000   
9997                     0                       0      5  1347494400   
9998                     1                       2      5  1203638400   
9999                     0                       1      4  1337472000   

                               Summary  \
0                Good Quality Dog Food   
1                    Not as Advertised   
2                "Delight" says it all   
3                       Cough Medicine   
4                          Great taffy   
...                                ...   
9995                      constipation   
9996  Constipation Not A Problem if...   
9997                Love this formula!   
9998                   very convenient   
9999        The best weve tried so far   

                                                   Text  \
0     I have bought several of the Vitality canned d...   
1     Product arrived labeled as Jumbo Salted Peanut...   
2     This is a confection that has been around a fe...   
3     If you are looking for the secret ingredient i...   
4     Great taffy at a great price.  There was a wid...   
...                                                 ...   
9995  we switched from the advance similac to the or...   
9996  Like the bad reviews say, the organic formula ...   
9997  I wanted to solely breastfeed but was unable t...   
9998  i love the fact that i can get this delieved t...   
9999  We have a 7 week old... He had gas and constip...   

                                      Clean Punctuation  \
0     I have bought several of the Vitality canned d...   
1     Product arrived labeled as Jumbo Salted Peanut...   
2     This is a confection that has been around a fe...   
3     If you are looking for the secret ingredient i...   
4     Great taffy at a great price  There was a wide...   
...                                                 ...   
9995  we switched from the advance similac to the or...   
9996  Like the bad reviews say the organic formula c...   
9997  I wanted to solely breastfeed but was unable t...   
9998  i love the fact that i can get this delieved t...   
9999  We have a 7 week old He had gas and constipati...   

                                        Clean Lowercase  \
0     i have bought several of the vitality canned d...   
1     product arrived labeled as jumbo salted peanut...   
2     this is a confection that has been around a fe...   
3     if you are looking for the secret ingredient i...   
4

In [73]:
data.to_csv('cleaned_reviews.csv', index=False)

# Feature Extraction

In [74]:
import pandas as pd
import math
from collections import Counter
from sklearn.feature_extraction.text import TfidfVectorizer

data = pd.read_csv('cleaned_reviews.csv')

# Drop rows with missing values in the 'Text' column
data.dropna(subset=['Clean Lemmatized2'], inplace=True)

# Extract text data from the 'Text' column
text_data = data['Clean Lemmatized2'].values.astype('U')

# Tokenize the text data
documents = [doc.split() for doc in text_data]

# Compute TF for each document
def compute_tf(document):
    word_count = Counter(document)
    tf = {word: count/len(document) for word, count in word_count.items()}
    return tf

tf_data = [compute_tf(doc) for doc in documents]

# Compute IDF
def compute_idf(documents):
    N = len(documents)
    idf = {}
    all_words = set(word for doc in documents for word in doc)
    for word in all_words:
        count = sum(1 for doc in documents if word in doc)
        idf[word] = math.log(N/count)
    return idf

idf = compute_idf(documents)

# Compute TF-IDF for each document
def compute_tfidf(document, idf):
    tfidf = {}
    tf = compute_tf(document)
    for word, tf_value in tf.items():
        tfidf[word] = tf_value * idf[word]
    return tfidf

tfidf_data = [compute_tfidf(doc, idf) for doc in documents]

# Convert TF-IDF data into DataFrame
tfidf_df = pd.DataFrame(tfidf_data).fillna(0)
print("\nTF-IDF Scores:")
print(tfidf_df)

# Using sklearn's TfidfVectorizer for TF-IDF computation
tfidf_vectorizer = TfidfVectorizer()
tfidf_matrix = tfidf_vectorizer.fit_transform(text_data)

# Convert TF-IDF matrix into DataFrame
tfidf_df_sklearn = pd.DataFrame(tfidf_matrix.toarray(), columns=tfidf_vectorizer.get_feature_names_out())
print("\nTF-IDF Scores using sklearn's TfidfVectorizer:")
print(tfidf_df_sklearn)



TF-IDF Scores:
      ['bought',  'sever',  'vital',    'can',    'dog',   'food',  \
0        0.15577  0.145881  0.310039  0.157201  0.134254  0.098157   
1        0.00000  0.000000  0.000000  0.000000  0.000000  0.000000   
2        0.00000  0.000000  0.000000  0.000000  0.000000  0.000000   
3        0.00000  0.000000  0.000000  0.000000  0.000000  0.000000   
4        0.00000  0.000000  0.000000  0.000000  0.000000  0.000000   
...          ...       ...       ...       ...       ...       ...   
9995     0.00000  0.000000  0.000000  0.000000  0.000000  0.056440   
9996     0.00000  0.000000  0.000000  0.000000  0.000000  0.000000   
9997     0.00000  0.000000  0.000000  0.000000  0.000000  0.000000   
9998     0.00000  0.000000  0.000000  0.000000  0.000000  0.000000   
9999     0.00000  0.000000  0.000000  0.000000  0.000000  0.000000   

      'product',  'found',   'good',  'qualiti',  ...  'bias']  'sucrosewel',  \
0       0.223380  0.112159  0.060816    0.131484  ...      0.0

# Machine Learning-based Method

### i. Naive-Bayes

In [87]:
from sklearn.model_selection import train_test_split
from sklearn.naive_bayes import MultinomialNB
from sklearn.metrics import accuracy_score, classification_report

def map_score_to_sentiment(score):
    if score >= 3:
        return 'positive'
    elif score <= 2:
        return 'negative'
    else:
        return 'neutral'

data['Sentiment'] = data['Score'].apply(map_score_to_sentiment)

X = tfidf_matrix
y = data['Sentiment']

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

nb_classifier = MultinomialNB()
nb_classifier.fit(X_train, y_train)

y_pred = nb_classifier.predict(X_test)

accuracy = accuracy_score(y_test, y_pred)
print("Accuracy:", accuracy)
print("\nClassification Report:")
print(classification_report(y_test, y_pred))


Accuracy: 0.8395

Classification Report:
              precision    recall  f1-score   support

    negative       0.00      0.00      0.00       321
    positive       0.84      1.00      0.91      1679

    accuracy                           0.84      2000
   macro avg       0.42      0.50      0.46      2000
weighted avg       0.70      0.84      0.77      2000



/Users/zafirahzaini/anaconda3/lib/python3.11/site-packages/sklearn/metrics/_classification.py:1469: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/Users/zafirahzaini/anaconda3/lib/python3.11/site-packages/sklearn/metrics/_classification.py:1469: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/Users/zafirahzaini/anaconda3/lib/python3.11/site-packages/sklearn/metrics/_classification.py:1469: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))


### ii. SVM

In [88]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.svm import LinearSVC
from sklearn.naive_bayes import MultinomialNB
from sklearn.metrics import accuracy_score, classification_report

def get_sentiment_val(score):
    if score >= 3:
        return 'positive'
    elif score <= 2:
        return 'negative'
    else:
        return 'neutral'

# Load the dataset
data = pd.read_csv('cleaned_reviews.csv', nrows=10000)
data = data[['Text', 'Score']].dropna()

# Assign sentiment labels
data['Sentiment'] = data['Score'].apply(get_sentiment_val)

# Define X and y based on your dataset
X = data['Text']
y = data['Sentiment']

# Split data into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# TF-IDF vectorization
tfidf_vectorizer = TfidfVectorizer(max_features=5000)
X_train_tfidf = tfidf_vectorizer.fit_transform(X_train)
X_test_tfidf = tfidf_vectorizer.transform(X_test)

# Support Vector Machines (SVM)
svm_model = LinearSVC()
svm_model.fit(X_train_tfidf, y_train)
svm_predictions = svm_model.predict(X_test_tfidf)
svm_accuracy = accuracy_score(y_test, svm_predictions)
print("SVM Accuracy:", svm_accuracy)

# Classification report for SVM
svm_classification_report = classification_report(y_test, svm_predictions)
print("SVM Classification Report:")
print(svm_classification_report)


SVM Accuracy: 0.8965
SVM Classification Report:
              precision    recall  f1-score   support

    negative       0.75      0.53      0.62       321
    positive       0.92      0.97      0.94      1679

    accuracy                           0.90      2000
   macro avg       0.83      0.75      0.78      2000
weighted avg       0.89      0.90      0.89      2000



/Users/zafirahzaini/anaconda3/lib/python3.11/site-packages/sklearn/svm/_classes.py:32: FutureWarning: The default value of `dual` will change from `True` to `'auto'` in 1.5. Set the value of `dual` explicitly to suppress the warning.
  warnings.warn(


# Lexicon-based Method

### i. Textblob

In [93]:
import pandas as pd
from sklearn.metrics import classification_report
from textblob import TextBlob

def get_sentiment_val(score):
    if score >= 3:
        return 'positive'
    elif score <= 2:
        return 'negative'
    else:
        return 'neutral'

# Load the dataset
data = pd.read_csv("cleaned_reviews.csv", nrows=10000)

# Map scores to binary sentiment labels
data['Sentiment'] = data['Score'].apply(get_sentiment_val)

# Define a function to perform sentiment analysis using TextBlob
def textblob_sentiment(text):
    analysis = TextBlob(text)
    return 'positive' if analysis.sentiment.polarity > 0 else 'negative'

# Apply sentiment analysis using TextBlob
data['TextBlob_Predictions'] = data['Text'].apply(textblob_sentiment)

# Generate classification report
accuracy = accuracy_score(data['Sentiment'], data['TextBlob_Predictions'])
print("Accuracy:", accuracy)
report = classification_report(data['Sentiment'], data['TextBlob_Predictions'])
print("\nClassification Report for Sentiment Analysis using TextBlob:\n", report)


Accuracy: 0.8553

Classification Report for Sentiment Analysis using TextBlob:
               precision    recall  f1-score   support

    negative       0.53      0.40      0.45      1522
    positive       0.90      0.94      0.92      8478

    accuracy                           0.86     10000
   macro avg       0.71      0.67      0.69     10000
weighted avg       0.84      0.86      0.85     10000



### ii. VADER

In [91]:
import pandas as pd
from sklearn.metrics import classification_report
from nltk.sentiment.vader import SentimentIntensityAnalyzer


def get_sentiment_val(score):
    if score >= 3:
        return 'positive'
    elif score <= 2:
        return 'negative'
    else:
        return 'neutral'

# Load the dataset
data = pd.read_csv("cleaned_reviews.csv")

# Map scores to binary sentiment labels
data['Sentiment'] = data['Score'].apply(get_sentiment_val)

# Initialize the VADER sentiment analyzer
analyzer = SentimentIntensityAnalyzer()

# Define a function to perform sentiment analysis using VADER
def vader_sentiment(text):
    compound_score = analyzer.polarity_scores(text)['compound']
    return 'positive' if compound_score > 0 else 'negative'

# Apply sentiment analysis using VADER
data['VADER_Predictions'] = data['Text'].apply(vader_sentiment)

# Generate classification report
accuracy = accuracy_score(data['Sentiment'], data['VADER_Predictions'])
print("Accuracy:", accuracy)
report = classification_report(data['Sentiment'], data['VADER_Predictions'])
print("\nClassification Report for Sentiment Analysis using VADER:\n", report)


Accuracy: 0.8582

Classification Report for Sentiment Analysis using VADER:
               precision    recall  f1-score   support

    negative       0.54      0.43      0.48      1522
    positive       0.90      0.93      0.92      8478

    accuracy                           0.86     10000
   macro avg       0.72      0.68      0.70     10000
weighted avg       0.85      0.86      0.85     10000



# Strength and Weakness

In the sentiment analysis, both machine learning-based approach and lexicon-based approach are observed for model selection. The strengths and weaknesses are observed based on the classification report. Based on the classification reports provided:

### Support Vector Machine (SVM):
- **Strengths**:
  - **High Accuracy**: Highest accuracy of 89.65% among the models selected.
  - **High Precision and Recall for Positive Class**: Highest recall (97%) and precision (92%) for the positive class as it can recognise pleasant emotions most of the time.
- **Weaknesses**:
  - **Lower Precision and Recall for Negative Class**: Indicates that it has greater difficulty distinguishing negative sentiments than positive sentiments, as evidenced by its poorer recall (53%) and precision (75%) for the negative class.

### Naive Bayes:
- **Strengths**:
  - **High Accuracy**: Achieves an accuracy of 83.95% which is still high in accuracy but not the best.
  - **High Precision and Recall for Positive Class**: Attains excellent recall (100%) and precision (84%) for the positive class.
- **Weaknesses**:
  - **Poor Performance for Negative Class**: Demonstrates low recall (0%), low precision (0%), and an inability to accurately identify negative thoughts for the negative class.

### TextBlob and VADER:
- **Strengths**:
  - **Balanced Precision and Recall**: TextBlob and VADER demonstrate strong performance of balanced precision and recall for both positive and negative classes.
- **Weaknesses**:
  - **Lower Overall Accuracy**: Accuracy is slightly lower than that of SVM and Naive Bayes.
  - **Lower F1-score for Negative Class**: Both TextBlob and VADER show Lower F1-scores for the negative class than the positive class, suggesting that it may be challenging to accurately identify negative thoughts.


In conclusion, SVM performs the best overall in terms of accuracy, precision, and recall for positive sentiment. TextBlob and VADER offer more balanced performance across classes but with slightly lower accuracy. Naive Bayes performs well for positive sentiments but poorly for negative sentiments.